docs/source/developers/python.rst

```diff
@@ -636,6 +636,52 @@ Caveats
 
 The Plasma component is not supported on Windows.
 
+Deleting stale build artifacts
+==============================
+
+When there have been changes to the structure of the Arrow C++ library or Pyarrow,
+a thorough cleaning is recommended as a first attempt to fixing build errors.
+
+.. note::
+
+   It is not necessarily intuitive from the error itself that the problem is due to stale artifacts.
+   Example of a build error from stale artifacts is "Unknown CMake command "arrow_keep_backward_compatibility".
+
+To delete stale Arrow C++ build artifacts:
+
+.. code-block::
+
+  $ rm -rf arrow/cpp/build
+
+To delete stale PyArrow build artifacts:
+
+.. code-block::
+
+  $ pushd python
+  $ python3 setup.py clean
```

Comment:
Now you mention to use both `python setup.py clean` and `git clean -Xfd .` from within the python directory. I think only mentioning the git way is sufficient? 
Higher up in this file, we also only mention git clean (this was changed from setup.py to git clean in https://github.com/apache/arrow/pull/11505). And additionally, it's actually deprecated to call setup.py directly like that (https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html, although we still use it for `python setup.py build_ext` anyway .. but we can at least not mention it for cleaning)

Reply:
One small follow-up comment: I would keep the `pushd python` when doing `git clean`, so it only cleans the python build artifacts, and not everything else (eg at this points in the explanation, you could first rebuild arrow cpp after cleaning that up, and then at that point you don't want to remove that again before doing a clean rebuild of pyarrow)

Or I suppose doing `git clean -Xdf python` works as well, instead of pushd/popd to python